### PR TITLE
Client iobase threadsafe

### PR DESCRIPTION
**Pull request by Jack-McDowell**

### CODE REVIEW

**Comment by Jack-McDowell** [comment]
Merge registry event monitor first

File: BLUESPAWN-common/headers/common/wrappers.hpp [hunk 1]
```diff
@@ -84,6 +84,14 @@ class FindWrapper : public GenericWrapper<HANDLE> {
 		GenericWrapper(handle, std::function<void(HANDLE)>(FindClose), INVALID_HANDLE_VALUE){};
 };
 
+class AcquireMutex : public GenericWrapper<HANDLE> {
+public:
+	AcquireMutex(HANDLE hMutex) :
```
Comment (ION28) [inline BLUESPAWN-common/headers/common/wrappers.hpp]: Codacy Issue found: [Class 'FindWrapper' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp [hunk 2]
```diff
@@ -7,82 +7,132 @@
...
-		return 1;
-	}
-	else if (negativeOptions.find(result) != negativeOptions.end()) {
-		return 0;
-	}
-	else {
-		return -1;
+
+DWORD CLI::GetUserConfirm(const std::wstring& prompt, DWORD dwMaximumDelay, ImportanceLevel level) const {
+	auto mutex = AcquireMutex(hMutex);
```
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [Variable 'mutex' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp [hunk 3]
```diff
@@ -7,82 +7,132 @@
...
-		wcin >> userIn;
-		if (userIn > 0 && userIn <= options.size()) {
-			return options[userIn-1];
-		}
-		else {
-			wcout << L"Please Enter a valid number between 1 and " << options.size() <<endl;
+	Print(L"Please enter a number 1 through " + std::to_wstring(i) + L" to continue. ", MessageColor::LIGHTGRAY, false);
+
+	while(true){
+		int userIn = 0;
```
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [The scope of the variable 'userIn' can be reduced.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/monitor/Event.cpp [hunk 4]
```diff
@@ -1,44 +1,182 @@
...
-	this->eventSub = std::move(sub);
-}
-
-std::wstring EventLogEvent::getChannel() {
+std::wstring EventLogEvent::GetChannel() const {
 	return channel;
 }
-int EventLogEvent::getEventID() {
+
+int EventLogEvent::GetEventID() const {
```
Comment (ION28) [inline BLUESPAWN-client/src/monitor/Event.cpp]: Codacy Issue found: [Variable 'thread' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/util/eventlogs/EventLogs.cpp [hunk 5]
```diff
@@ -9,291 +9,196 @@ const int ARRAY_SIZE = 10;
...
 		// Provide values for filtered parameters
 		for (std::wstring key : params) {
-			std::wstring val;
-			if (ERROR_SUCCESS != (status = GetEventParam(hEvent, &val, key))) {
-				LOG_ERROR(L"EventLogs::EventToDetection: Failed query parameter " + key + L" with code " + std::to_wstring(status));
-				return status;
+			std::optional<std::wstring> val = GetEventParam(hEvent, key);
+			if (!val) {
+				return std::nullopt;
 			}
```
Comment (ION28) [inline BLUESPAWN-client/src/util/eventlogs/EventLogs.cpp]: Codacy Issue found: [Variable 'status' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-common/headers/common/wrappers.hpp [hunk 6]
```diff
@@ -84,6 +84,14 @@ class FindWrapper : public GenericWrapper<HANDLE> {
...
+		GenericWrapper(hMutex, std::function<void(HANDLE)>(ReleaseMutex), INVALID_HANDLE_VALUE){
+		WaitForSingleObject(hMutex, INFINITE);
+	};
+};
+
 class AllocationWrapper {
```
Comment (ION28) [inline BLUESPAWN-common/headers/common/wrappers.hpp]: Codacy Issue found: [Class 'AcquireMutex' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp (hunk 3 above)
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [The scope of the variable 'userIn' can be reduced.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp (hunk 2 above)
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [Variable 'mutex' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-common/headers/common/wrappers.hpp (hunk 1 above)
Comment (ION28) [inline BLUESPAWN-common/headers/common/wrappers.hpp]: Codacy Issue found: [Class 'AcquireMutex' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp (hunk 3 above)
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [The scope of the variable 'userIn' can be reduced.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp (hunk 2 above)
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [Variable 'mutex' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-common/headers/common/wrappers.hpp (hunk 1 above)
Comment (ION28) [inline BLUESPAWN-common/headers/common/wrappers.hpp]: Codacy Issue found: [Class 'AcquireMutex' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp [hunk 7]
```diff
@@ -7,82 +7,136 @@
...
```
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [The scope of the variable 'userIn' can be reduced.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/user/CLI.cpp [hunk 8]
```diff
@@ -7,82 +7,136 @@
...
-		return 1;
-	}
-	else if (negativeOptions.find(result) != negativeOptions.end()) {
-		return 0;
-	}
-	else {
-		return -1;
+
+DWORD CLI::GetUserConfirm(const std::wstring& prompt, DWORD dwMaximumDelay, ImportanceLevel level) const {
+	auto mutex = AcquireMutex(hMutex);
```
Comment (ION28) [inline BLUESPAWN-client/src/user/CLI.cpp]: Codacy Issue found: [Variable 'mutex' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

File: BLUESPAWN-client/src/util/log/CLISink.cpp [hunk 9]
```diff
@@ -11,7 +11,10 @@ namespace Log {
 		SetConsoleTextAttribute(hConsole, static_cast<WORD>(color));
 	}
 
+	CLISink::CLISink(const HandleWrapper& hMutex) : hMutex{ hMutex } {}
+
 	void CLISink::LogMessage(const LogLevel& level, const std::string& message, const std::optional<HuntInfo> info, const std::vector<std::shared_ptr<DETECTION>>& detections){
+		auto mutex = AcquireMutex(hMutex);
```
Comment (ION28) [inline BLUESPAWN-client/src/util/log/CLISink.cpp]: Codacy Issue found: [Variable 'mutex' is assigned a value that is never used.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=5011005)

**Pull request merged**